docs: add new blogpost about security in cilium (#965)
cilium/cilium.io: src/posts/2026-05-06-securing-cicd-open-source-lessons-from-cilium/index.md

Cilium runs in the kernel-level networking path of millions of Kubernetes pods. If our supply chain were compromised, the blast radius would not be small. Hardening the project against that scenario is something we work on continuously, and we wanted to write down what we actually do, in detail. Most of what follows isn't Cilium-specific: any open source project running CI/CD on GitHub Actions can apply these patterns. We've also called out where we still fall short, in case any of it makes a useful starting point for someone else.

## TL;DR

If you don't have time to read the whole thing, here's what Cilium does to harden its supply chain today:

> 

Reviewer: one idea could be to turn this into a table to make it more easily scannable and make a first column about the layer to highlight the different layers of defense in depth. Also fine as is, just an idea :)

Author: Yep, it does look better with a table.

[`.github/workflows/build-images-ci.yaml`](https://github.com/cilium/cilium/blob/main/.github/workflows/build-images-ci.yaml)

```yaml
- name: Checkout base or default branch (trusted)
  uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
  with:
    ref: ${{ github.base_ref || github.event.repository.default_branch }}
    persist-credentials: false

# ...trusted setup steps run here, including loading composite actions...

# Warning: since this is a privileged workflow, subsequent workflow job
# steps must take care not to execute untrusted code.
- name: Checkout pull request branch (NOT TRUSTED)
  uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
  with:
    persist-credentials: false
    ref: ${{ steps.tag.outputs.sha }}
```

The first checkout grabs the _base branch_ (code that's already been reviewed and merged) so we can load our composite actions, scripts, and the Cosign signing logic from a known-good source. Only after that does the workflow check out the PR head, and that checkout is used purely as build context for `docker build`. Nothing from the PR branch is ever executed as a script.

> We get security reports about this pattern fairly regularly. Automated scanners and well-meaning researchers see "`pull_request_target` plus a second checkout" and flag it as a vulnerability. In the general case they're right to. In ours the workflow is intentionally designed so the pattern is safe:

Reviewer: ```suggestion
We get security reports about this pattern fairly regularly. Automated scanners and well-meaning researchers see "`pull_request_target` plus a second checkout" and flag it as a vulnerability. In the general case they're right too. In ours, the workflow is intentionally designed so the pattern is safe:
```

[`.github/renovate.json5`](https://github.com/cilium/cilium/blob/main/.github/renovate.json5)

```json5
{
  // Dependency cooldown: skip versions published less than 5 days ago
  "matchUpdateTypes": ["major", "minor", "patch"],
  "minimumReleaseAge": "5 days"
},
{
  "matchPackageNames": [
    "actions/{/,}**",                  // GitHub's official actions
    "docker/{/,}**",                   // Official Docker actions
    "cilium/{/,}**",                   // Our own ecosystem
    "k8s.io/{/,}**",                   // Kubernetes official
    "sigs.k8s.io/{/,}**",              // Kubernetes SIGs
    "golang.org/x/{/,}**",             // Go experimental
    "github.com/golang/{/,}**",        // Go official org
    "github.com/prometheus/{/,}**",
    "github.com/hashicorp/{/,}**",
    "go.etcd.io/etcd/{/,}**",
    // ...trimmed
  ],
  "automerge": true,
  "automergeType": "pr",
  "groupName": "auto-merge-trusted-deps",
  "reviewers": ["ciliumbot"]
}
```

Updates from this allow-list auto-merge after CI passes. Everything else needs a human review.

> The [auto-approve workflow](https://github.com/cilium/cilium/blob/main/.github/workflows/auto-approve.yaml) adds another belt-and-braces check: it verifies that the PR was created by `cilium-renovate[bot]` _and_ that the review request was actually triggered by the bot itself, not by a human pretending to be it:

Reviewer: ```suggestion
The [auto-approve workflow](https://github.com/cilium/cilium/blob/main/.github/workflows/auto-approve.yaml) adds another belt-and-suspenders check: it verifies that the PR was created by `cilium-renovate[bot]` _and_ that the review request was actually triggered by the bot itself, not by a human pretending to be it:
```

Release builds also run inside [protected environments](https://docs.github.com/en/actions/deployment/targeting-different-environments/managing-environments-for-deployment) (`release`, `release-tool`, `release-helm`) so production registry credentials are gated behind environment protection rules. You can't trigger a release build from a fork or a feature branch.

## The Cilium security team

If you've ever reported a security issue to the project (via [GitHub security advisories](https://github.com/cilium/cilium/security/advisories) or [security@cilium.org](mailto:security@cilium.org)), you've already interacted with [Cilium's Security Team](https://github.com/cilium/community/blob/main/roles/Security-Team.md). Beyond triaging vulnerability reports, the team also runs the operational side of supply chain security:

- Auditing and rotating credentials and permissions across the GitHub organization.
- When necessary, carrying out incident investigation and audits.
- Monitoring for patterns in our security issues and industry developments in order to propose mitigations and controls in areas where our security posture is weak.

> ## Additional layers

Reviewer: is it worth mentioning the fuzzing we did here?

Author: I feel like fuzzing hasn't been that useful to put in here.

- **No `govulncheck` in CI.** We fuzz and we lint, but we don't yet run Go's official vulnerability scanner, which checks whether our code actually calls vulnerable functions rather than just whether a vulnerable package shows up in `go.sum`.
- **68 internal `@main` references.** A bunch of conformance and scale-test workflows reference `cilium/cilium/.github/actions/set-commit-status@main`, which is a mutable branch ref. It's lower risk than a third-party tag, but inconsistent with our SHA-pinning policy. The plan is to move all of our composite actions out of cilium/cilium into a dedicated repository, which removes the need for `@main` here.

A few smaller items in the same audit:

- No [OpenSSF Scorecard](https://securityscorecards.dev/) workflow for continuous supply chain health monitoring.
- Our `SECURITY-INSIGHTS.yml` expired in January 2025 and hasn't been updated. (We actually noticed this while writing this post.)
- No `go mod verify` step to validate vendor directory integrity against `go.sum` checksums.

If any of these look like a good first issue and you want to send a PR, we'd take it.

> ## GitHub's 2026 Actions security roadmap and how it maps to what we do

Reviewer: maybe add a divider here since it is transitioning away from the project now?

### Scoped secrets: closing the implicit inheritance gap

CI vs. production credential isolation is one of our strongest controls, but within a given environment, secrets are still scoped pretty broadly: any workflow running in that environment can access them.

Scoped secrets would let us bind credentials to specific workflow paths, branches, or even individual reusable workflows. A release credential could be restricted not just to the `release` environment but to the specific `release.yaml` workflow file, so a new workflow added to that environment (by accident or by an attacker) wouldn't inherit the credentials. That's a meaningful step beyond what protected environments alone provide.

The roadmap also separates secret management from repository write access. Today anyone with write access to a repo can manage its secrets. GitHub plans to move secret management into a dedicated custom role, which lines up with the least-privilege principle we already apply to workflow permissions but can't currently apply to secret administration.

### Native egress firewall

> GitHub's planned native egress firewall would restrict outbound network access from GitHub-hosted runners. It runs outside the runner VM at L7, so it's immutable even if an attacker gets root inside the runner. Organisations would define allowed domains, IP ranges, and HTTP methods, and anything else gets blocked.

Reviewer: ```suggestion
GitHub's planned native egress firewall would restrict outbound network access from GitHub-hosted runners. It runs outside the runner VM at L7, so it's immutable even if an attacker gets root inside the runner. Organizations would define allowed domains, IP ranges, and HTTP methods, and anything else gets blocked.
```

> For Cilium it's less critical than the rest. Our most security-sensitive workflows (release builds, image signing) already run with credential isolation and least-privilege permissions, which limits what a compromised step could do even with unrestricted network access. Building an accurate egress allowlist for a project that talks to container registries, Go module proxies, cloud APIs, and Sigstore would be a significant chunk of work. Public preview is expected in 6 to 9 months, so we'll evaluate then.

Reviewer: ```suggestion
For Cilium it's less critical than the rest. Our most security-sensitive workflows (release builds, image signing) already run with credential isolation and least-privilege permissions, which limits what a compromised step could do even with unrestricted network access. Building an accurate egress allow-list for a project that talks to container registries, Go module proxies, cloud APIs, and Sigstore would be a significant chunk of work. Public preview is expected in 6 to 9 months, so we'll evaluate then.
```
